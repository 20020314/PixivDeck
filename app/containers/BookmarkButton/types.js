// @flow
export type Restrict = 'public' | 'private'
