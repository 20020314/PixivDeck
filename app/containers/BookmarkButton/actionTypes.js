// @flow
import type { Restrict } from './types'

export type ADD_BOOKMARK_REQUEST_TYPE = 'BookmarkButton/ADD_BOOKMARK_REQUEST'
export type ADD_BOOKMARK_SUCCESS_TYPE = 'BookmarkButton/ADD_BOOKMARK_SUCCESS'
export type ADD_BOOKMARK_FAILER_TYPE = 'BookmarkButton/ADD_BOOKMARK_FAILER'

export type Action =
  | {|
      +type: ADD_BOOKMARK_REQUEST_TYPE | ADD_BOOKMARK_SUCCESS_TYPE,
      +id: number,
      +restrict: Restrict,
    |}
  | {| +type: ADD_BOOKMARK_FAILER_TYPE, +id: number, +error: string |}
