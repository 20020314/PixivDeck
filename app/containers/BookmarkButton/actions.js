// @flow
import type { Restrict } from './types'

import {
  ADD_BOOKMARK_REQUEST,
  ADD_BOOKMARK_SUCCESS,
  ADD_BOOKMARK_FAILURE,
  DELETE_BOOKMARK_REQUEST,
  DELETE_BOOKMARK_SUCCESS,
  DELETE_BOOKMARK_FAILURE,
} from './actionTypes'
import type {
  AddBookmarkRequest,
  AddBookmarkSuccess,
  AddBookmarkFailure,
  DeleteBookmarkRequest,
  DeleteBookmarkSuccess,
  DeleteBookmarkFailure,
} from './actionTypes'

export function addBookmarkRequest(
  id: number,
  restrict: Restrict
): AddBookmarkRequest {
  return {
    type: ADD_BOOKMARK_REQUEST,
    id,
    restrict,
  }
}
export function addBookmarkSuccess(
  id: number,
  restrict: Restrict
): AddBookmarkSuccess {
  return {
    type: ADD_BOOKMARK_SUCCESS,
    id,
    restrict,
  }
}
export function addBookmarkFailure(
  id: number,
  error: string
): AddBookmarkFailure {
  return {
    type: ADD_BOOKMARK_FAILURE,
    id,
    error,
  }
}
export function deleteBookmarkRequest(id: number): DeleteBookmarkRequest {
  return {
    type: DELETE_BOOKMARK_REQUEST,
    id,
  }
}
export function deleteBookmarkSuccess(id: number): DeleteBookmarkSuccess {
  return {
    type: DELETE_BOOKMARK_SUCCESS,
    id,
  }
}
export function deleteBookmarkFailure(
  id: number,
  error: string
): DeleteBookmarkFailure {
  return {
    type: DELETE_BOOKMARK_FAILURE,
    id,
    error,
  }
}
