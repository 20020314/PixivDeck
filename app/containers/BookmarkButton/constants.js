// @flow
import type {
  ADD_BOOKMARK_REQUEST_TYPE,
  ADD_BOOKMARK_SUCCESS_TYPE,
  ADD_BOOKMARK_FAILER_TYPE,
} from './actionTypes.js'

export const ADD_BOOKMARK_REQUEST: ADD_BOOKMARK_REQUEST_TYPE =
  'BookmarkButton/ADD_BOOKMARK_REQUEST'
export const ADD_BOOKMARK_SUCCESS: ADD_BOOKMARK_SUCCESS_TYPE =
  'BookmarkButton/ADD_BOOKMARK_SUCCESS'
export const ADD_BOOKMARK_FAILER: ADD_BOOKMARK_FAILER_TYPE =
  'BookmarkButton/ADD_BOOKMARK_FAILER'
