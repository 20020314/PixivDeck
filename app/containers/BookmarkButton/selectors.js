// @flow
import { createSelector } from 'reselect'
import type { State } from 'types/state'

const getSelectIllust = (state: State, { id }: { id: number }) =>
  state.illustById[id]

export const makeSelectIllust = () => createSelector(getSelectIllust, s => s)

export const makeIsBookmarked = () =>
  createSelector(makeSelectIllust(), s => s.isBookmarked)
