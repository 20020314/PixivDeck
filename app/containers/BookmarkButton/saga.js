// @flow
import { call, put, takeEvery, type IOEffect } from 'redux-saga/effects'
import { get } from '../Api/sagas'
import * as api from 'containers/Api/sagas'
import * as Actions from './constants'
import * as actions from './actions'
import type { Restrict } from './types'

type Props = {
  id: number,
  restrict: Restrict,
}

export function* bookmark({ id, restrict }: Props): Generator<*, void, void> {
  try {
    yield call(api.post, '/v2/illust/bookmark/add', { illustId: id, restrict })

    yield put(actions.addBookmarkSuccess(id, restrict))
  } catch (err) {
    yield put(actions.addBookmarkFailer(id, err))
  }
}

export function* deleteTask({ id }: Props): Generator<*, void, void> {
  try {
    yield call(api.post, '/v1/illust/bookmark/delete', { illustId: id })
    yield put(actions.deleteBookmarkSuccess(id))
  } catch (err) {
    yield put(actions.deleteBookmarkFailer(id, err))
  }
}

function* success({ id }: Props) {
  try {
    yield call(get, `/v1/illust/detail?illust_id=${id}`, true)
  } catch (err) {}
}

function* root(): Generator<IOEffect, void, *> {
  yield takeEvery(Actions.ADD_BOOKMARK_REQUEST, bookmark)
  yield takeEvery(Actions.DELETE_BOOKMARK_REQUEST, deleteTask)
  yield takeEvery(
    [Actions.ADD_BOOKMARK_SUCCESS, Actions.DELETE_BOOKMARK_SUCCESS],
    success
  )
}

export default root
