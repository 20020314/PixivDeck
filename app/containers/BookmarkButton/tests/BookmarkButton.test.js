// @flow
import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'
import BookmarkButton from '../BookmarkButton'

const onClick = jest.fn()

const setup = props => {
  return {
    onClick,
    isBookmarked: false,
    addBookmark: jest.fn(),
    deleteBookmark: jest.fn(),
    ...props,
  }
}

test('snapshot BookmarkButton isBookmarked=true', () => {
  const comp = shallow(<BookmarkButton {...setup({ isBookmarked: true })} />)
  expect(toJSON(comp)).toMatchSnapshot()
})

test('snapshot BookmarkButton isBookmarked=false', () => {
  const comp = shallow(<BookmarkButton {...setup({ isBookmarked: false })} />)
  expect(toJSON(comp)).toMatchSnapshot()
})
