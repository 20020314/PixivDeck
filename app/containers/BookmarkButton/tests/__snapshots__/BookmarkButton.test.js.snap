// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot BookmarkButton isBookmarked=false 1`] = `
<IconButton
  disableTouchRipple={false}
  disabled={false}
  iconStyle={
    Object {
      "height": 18,
      "width": 18,
    }
  }
  onClick={[MockFunction]}
  style={
    Object {
      "height": 32,
      "width": 32,
    }
  }
  tooltipPosition="bottom-center"
  touch={false}
>
  <ActionFavorite
    color="#777"
    hoverColor="#b94343"
  />
</IconButton>
`;

exports[`snapshot BookmarkButton isBookmarked=true 1`] = `
<IconButton
  disableTouchRipple={false}
  disabled={false}
  iconStyle={
    Object {
      "height": 18,
      "width": 18,
    }
  }
  onClick={[MockFunction]}
  style={
    Object {
      "height": 32,
      "width": 32,
    }
  }
  tooltipPosition="bottom-center"
  touch={false}
>
  <ActionFavorite
    color="#b94343"
    hoverColor="#b94343"
  />
</IconButton>
`;
